pkg/specs/content.go

```diff
@@ -0,0 +1,227 @@
+	"github.com/mitchellh/cli"
+	"github.com/pkg/errors"
+	"github.com/replicatedhq/ship/pkg/api"
+	"github.com/replicatedhq/ship/pkg/constants"
+	"github.com/replicatedhq/ship/pkg/fs"
+	"github.com/replicatedhq/ship/pkg/specs/apptype"
+	"github.com/replicatedhq/ship/pkg/specs/githubclient"
+	"github.com/replicatedhq/ship/pkg/specs/replicatedapp"
+	"github.com/replicatedhq/ship/pkg/state"
+	"github.com/replicatedhq/ship/pkg/util"
+	"github.com/spf13/afero"
+	"github.com/spf13/viper"
+)
+
+type ContentProcessor struct {
+	Logger           log.Logger
+	FS               afero.Afero
+	AppResolver      replicatedapp.Resolver
+	GitHubFetcher    githubclient.GitHubFetcher
+	ui               cli.Ui
+	appTypeInspector apptype.Inspector
+	shaSummer        shaSummer
+	rootDir          string
+}
+
+func NewContentProcessor(v *viper.Viper) (*ContentProcessor, error) {
+	logger := level.NewFilter(log.NewLogfmtLogger(os.Stdout), level.AllowInfo())
+	fs := fs.NewBaseFilesystem()
+
+	dir, err := fs.TempDir("", "watch")
+	if err != nil {
+		return nil, err
+	}
+	constants.SetShipRootDir(dir)
+
+	if err := os.MkdirAll(constants.ShipPathInternalTmp, 0755); err != nil {
+		return nil, err
+	}
+
+	ui := &cli.BasicUi{
+		Reader:      os.Stdin,
+		Writer:      ioutil.Discard,
+		ErrorWriter: os.Stderr,
+	}
+	gqlClient, err := replicatedapp.NewGraphqlClient(v, http.DefaultClient)
+	if err != nil {
+		return nil, err
+	}
+	stateManager := state.NewManager(logger, fs, v)
+	appTypeInspector := apptype.NewInspector(logger, fs, v, stateManager, ui)
+	return &ContentProcessor{
+		Logger:           logger,
+		FS:               fs,
+		AppResolver:      replicatedapp.NewAppResolver(v, logger, fs, gqlClient, stateManager, ui),
+		GitHubFetcher:    githubclient.NewGithubClient(logger),
+		ui:               ui,
+		appTypeInspector: appTypeInspector,
+		shaSummer:        calculateContentSHA,
+		rootDir:          dir,
+	}, nil
+}
+
+func (r *Resolver) NewContentProcessor() *ContentProcessor {
+	return &ContentProcessor{
+		Logger:           r.Logger,
+		FS:               r.FS,
+		AppResolver:      r.AppResolver,
+		GitHubFetcher:    r.GitHubFetcher,
+		ui:               r.ui,
+		appTypeInspector: r.appTypeInspector,
+		shaSummer:        r.shaSummer,
+	}
+}
+
+// MaybeResolveVersionedUpstream returns an upstream with the UpstreamVersionToken replaced with the
+// latest version fetched from github unless the latest version is unable to be fetched or if the version
+// stored in state is greater than the latest version. All other upstreams will be returned unmodified.
+func (c *ContentProcessor) MaybeResolveVersionedUpstream(ctx context.Context, upstream string, existingState state.State) (string, error) {
+	debug := level.Debug(log.With(c.Logger, "method", "resolveVersionedUpstream"))
+
+	if !util.IsGithubURL(upstream) {
+		return upstream, nil
+	}
+
+	debug.Log("event", "resolve latest release")
+	latestReleaseVersion, err := c.GitHubFetcher.ResolveLatestRelease(ctx, upstream)
```

Comment:
I know this would be a logic change, but perhaps we should skip checking for a newer release via github if the persisted URL doesn't contain the UpstreamVersionToken?

Reply:
This should be a seperate PR if it ends up being desirable